packages/application-generic/src/usecases/select-variant/select-variant.usecase.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable, Logger } from '@nestjs/common';
+
+import { MessageTemplateRepository, MessageTemplateEntity } from '@novu/dal';
+import { ICondition } from '@novu/shared';
+
+import { SelectVariantCommand } from './select-variant.command';
+import { ConditionsFilter } from '../conditions-filter';
+import { ConditionsFilterCommand } from '../conditions-filter';
+import { PlatformException } from '../../utils/exceptions';
+import { IFilterVariables } from '../../utils/filter-processing-details';
+
+const LOG_CONTEXT = 'SelectVariant';
+
+@Injectable()
+export class SelectVariant {
+  constructor(
+    private conditionsFilter: ConditionsFilter,
+    private messageTemplateRepository: MessageTemplateRepository
+  ) {}
+
+  async execute(command: SelectVariantCommand): Promise<{
+    messageTemplate: MessageTemplateEntity;
+    conditions?: ICondition[];
+  }> {
+    if (!command.step.variants?.length) {
+      return { messageTemplate: command.step.template };
+    }
+
+    if (!this.isFilterDataExist(command.filterData)) {
+      return { messageTemplate: command.step.template };
+    }
+
+    for (const variant of command.step.variants) {
+      if (!variant.filters?.length) {
+        continue;
+      }
+
+      const { passed, conditions } = await this.conditionsFilter.filter(
+        ConditionsFilterCommand.create({
+          filters: variant.filters,
+          environmentId: command.environmentId,
+          organizationId: command.organizationId,
+          userId: command.userId,
+          step: command.step,
+          job: command.job,
+        }),
+        command.filterData
+      );
+
+      if (passed) {
+        const messageTemplate = await this.messageTemplateRepository.findOne({
+          _organizationId: command.organizationId,
+          _environmentId: command.environmentId,
+          _id: variant._templateId,
+        });
+
+        if (!messageTemplate) {
+          const errorMessage = `Variant message template with id ${variant._templateId} not found`;
+          Logger.error(
+            {
+              variantTemplateId: variant._templateId,
+              filters: variant.filters,
+              organizationId: command.organizationId,
+              transactionId: command.job.transactionId,
+              conditions,
+            },
+            errorMessage,
+            LOG_CONTEXT
+          );
+
+          throw new PlatformException(errorMessage);
+        }
+
+        return { messageTemplate, conditions };
+      }
+    }
+
+    return { messageTemplate: command.step.template };
+  }
+
+  private isFilterDataExist(filterData: IFilterVariables) {
+    return (
+      !!filterData.tenant ||
+      !!filterData.payload ||
+      !!filterData.subscriber ||
+      !!filterData.webhook
+    );
```

Comment:
But what if the condition is using tenant fields but this return true because it has subscriber data?

Reply:
Should be totally fine that means that the filter will fail, but we could make this check more strict with some mapping i am just afraid it will make it more complex.

This one is a minor check in order to test if any data exists, initialy i created it when i thought it would be using only tenant data, so we could remove it as well.